### PR TITLE
Update candidate

### PR DESCRIPTION
**Pull request by voidsong-dragonfly**
Add in comparator support for hatches & Update hatch fluid handling code
Refactors from Bryce92

### CODE REVIEW

File: src/main/java/zmaster587/libVulpes/api/material/MaterialRegistry.java [hunk 1]
```diff
@@ -252,44 +253,47 @@ public Block getBlockForProduct(AllowedProducts product, zmaster587.libVulpes.ap
...
 		return getItemStackFromMaterialAndType(material, product,1);
 	}
 
 	/**
-	 * @param material
+	 * @param ore
 	 * @param product
 	 * @param amount stackSize
-	 * @return an itemstack of stackSize amount containing the product with the given material, or null if one does not exist
+	 * @return an EtemStack of stackSize amount containing the product with the given material, or ItemStack.EMPTY if one does not exist
```
Comment (zmaster587) [inline src/main/java/zmaster587/libVulpes/api/material/MaterialRegistry.java]:
```suggestion
	 * @return an ItemStack of stackSize amount containing the product with the given material, or ItemStack.EMPTY if one does not exist
```

File: src/main/java/zmaster587/libVulpes/client/util/ProgressBarImage.java [hunk 2]
```diff
@@ -150,10 +150,10 @@ public void drawTexturedModalRect(int x, int zLevel, int y, int textureX, int te
         Tessellator tessellator = Tessellator.getInstance();
         BufferBuilder vertexbuffer = tessellator.getBuffer();
         vertexbuffer.begin(7, DefaultVertexFormats.POSITION_TEX);
-        vertexbuffer.pos((double)(x + 0), (double)(y + height), (double)zLevel).tex((double)((float)(textureX + 0) * 0.00390625F), (double)((float)(textureY + height) * 0.00390625F)).endVertex();
-        vertexbuffer.pos((double)(x + width), (double)(y + height), (double)zLevel).tex((double)((float)(textureX + width) * 0.00390625F), (double)((float)(textureY + height) * 0.00390625F)).endVertex();
-        vertexbuffer.pos((double)(x + width), (double)(y + 0), (double)zLevel).tex((double)((float)(textureX + width) * 0.00390625F), (double)((float)(textureY + 0) * 0.00390625F)).endVertex();
-        vertexbuffer.pos((double)(x + 0), (double)(y + 0), (double)zLevel).tex((double)((float)(textureX + 0) * 0.00390625F), (double)((float)(textureY + 0) * 0.00390625F)).endVertex();
+        vertexbuffer.pos(x    , y + height, zLevel).tex((float)(textureX    ) * 0.00390625F, (float)(textureY + height) * 0.00390625F).endVertex();
```
Comment (zmaster587) [inline src/main/java/zmaster587/libVulpes/client/util/ProgressBarImage.java]:
```suggestion
        vertexbuffer.pos(x, y + height, zLevel).tex((float)(textureX) * 0.00390625F, (float)(textureY + height) * 0.00390625F).endVertex();
```

File: src/main/java/zmaster587/libVulpes/client/util/ProgressBarImage.java [hunk 3]
```diff
@@ -150,10 +150,10 @@ public void drawTexturedModalRect(int x, int zLevel, int y, int textureX, int te
...
+        vertexbuffer.pos(x + width, y + height, zLevel).tex((float)(textureX + width) * 0.00390625F, (float)(textureY + height) * 0.00390625F).endVertex();
+        vertexbuffer.pos(x + width, y    , zLevel).tex((float)(textureX + width) * 0.00390625F, (float)(textureY    ) * 0.00390625F).endVertex();
```
Comment (zmaster587) [inline src/main/java/zmaster587/libVulpes/client/util/ProgressBarImage.java]:
```suggestion
        vertexbuffer.pos(x + width, y, zLevel).tex((float)(textureX + width) * 0.00390625F, (float)(textureY) * 0.00390625F).endVertex();
```

File: src/main/java/zmaster587/libVulpes/client/util/ProgressBarImage.java [hunk 4]
```diff
@@ -150,10 +150,10 @@ public void drawTexturedModalRect(int x, int zLevel, int y, int textureX, int te
...
+        vertexbuffer.pos(x    , y    , zLevel).tex((float)(textureX    ) * 0.00390625F, (float)(textureY    ) * 0.00390625F).endVertex();
```
Comment (zmaster587) [inline src/main/java/zmaster587/libVulpes/client/util/ProgressBarImage.java]:
```suggestion
        vertexbuffer.pos(x, y, zLevel).tex((float)(textureX) * 0.00390625F, (float)(textureY) * 0.00390625F).endVertex();
```

File: src/main/java/zmaster587/libVulpes/inventory/ContainerModular.java [hunk 5]
```diff
@@ -136,9 +141,9 @@ else if (!this.mergeItemStack(itemstack1, 0, slotId, false))
 	}
 	
 	 /**
-     * merges provided ItemStack with the first avaliable one in the container/player inventory
+     * merges provided ItemStack with the first available one in the container/player inventory
      */
-    /*protected boolean mergeItemStack(ItemStack p_75135_1_, int p_75135_2_, int p_75135_3_, boolean p_75135_4_)
+    /*protected boolean mergeItemStack(@Nonnull ItemStack p_75135_1_, int p_75135_2_, int p_75135_3_, boolean p_75135_4_)
```
Comment (zmaster587) [inline src/main/java/zmaster587/libVulpes/inventory/ContainerModular.java]: This whole commented out bit can probably be removed.

File: src/main/java/zmaster587/libVulpes/items/ItemLinker.java [hunk 6]
```diff
@@ -91,40 +92,37 @@ public static void setDimId(ItemStack itemStack, int id) {
 		
 	}
 	
-	public static int getDimId(ItemStack itemStack) {
+	public static int getDimId(@Nonnull ItemStack itemStack) {
 		NBTTagCompound nbt;
-		if(!itemStack.hasTagCompound()) {
+		if(!itemStack.hasTagCompound() || (nbt = itemStack.getTagCompound()) == null)
```
Comment (zmaster587) [inline src/main/java/zmaster587/libVulpes/items/ItemLinker.java]: While technically correct, it's a lot less readable that using else

File: src/main/java/zmaster587/libVulpes/items/ItemMaterialBlock.java [hunk 7]
```diff
@@ -3,14 +3,16 @@
...
+
 public class ItemMaterialBlock extends ItemBlock {
 
 	public ItemMaterialBlock(Block block) {
 		super(block);
 	}
 
 	/*@Override
-	public boolean placeBlockAt(ItemStack stack, EntityPlayer player,
+	public boolean placeBlockAt(@Nonnull ItemStack stack, EntityPlayer player,
```
Comment (zmaster587) [inline src/main/java/zmaster587/libVulpes/items/ItemMaterialBlock.java]: We can probably remove these commented out methods.

File: src/main/java/zmaster587/libVulpes/tile/energy/TileCreativePowerInput.java [hunk 8]
```diff
@@ -63,14 +65,17 @@ public String getModularInventoryName() {
...
 	public String getName() {
 		// TODO Auto-generated method stub
 		return null;
 	}
 
 	@Override
 	public int receiveEnergy(int amt, boolean simulate) {
-		return receiveEnergy(amt, simulate);
+		return 0;
+		// TODO why was this infinite recursion here?
```
Comment (zmaster587) [inline src/main/java/zmaster587/libVulpes/tile/energy/TileCreativePowerInput.java]: Not sure, may have been pointing to a method with a different sig at some point.

File: src/main/java/zmaster587/libVulpes/tile/multiblock/TileMultiBlock.java [hunk 9]
```diff
@@ -200,12 +198,15 @@ public void deconstructMultiBlock(World world, BlockPos destroyedPos, boolean bl
 					int globalZ = pos.getZ() - (x - offset.x)*front.getFrontOffsetX()  - (z-offset.z)*front.getFrontOffsetZ();
 
 
+					TileEntity tile = world.getTileEntity(new BlockPos(globalX, globalY, globalZ));
 					//This block is being broken anyway so don't bother
-					if(blockBroken && globalX == destroyedPos.getX() &&
-							globalY == destroyedPos.getY() &&
-							globalZ == destroyedPos.getZ())
+					if(blockBroken && globalX == destroyedPos.getX() && globalY == destroyedPos.getY() && globalZ == destroyedPos.getZ()) {
```
Comment (zmaster587) [inline src/main/java/zmaster587/libVulpes/tile/multiblock/TileMultiBlock.java]: Should stay on multiple lines for readability, however the operator should come first, not at the end like the previous code did.

```suggestion
					if(blockBroken 
					&& globalX == destroyedPos.getX() 
					&& globalY == destroyedPos.getY() 
					&& globalZ == destroyedPos.getZ()) {
```

File: src/main/java/zmaster587/libVulpes/tile/multiblock/TileMultiBlock.java [hunk 10]
```diff
@@ -366,10 +367,10 @@ else if(((IMultiblock)tile).getMasterBlock() == this)
 							continue;
 					}
 					//Make sure the structure is valid
-					if(!(structure[y][z][x] instanceof Character && (Character)structure[y][z][x] == 'c') && !(structure[y][z][x] instanceof Block && (Block)structure[y][z][x] == Blocks.AIR && world.isAirBlock(globalPos)) && !getAllowableBlocks(structure[y][z][x]).contains(new BlockMeta(block,meta))) {
+					if(!(structure[y][z][x] instanceof Character && (Character)structure[y][z][x] == 'c') && !(structure[y][z][x] instanceof Block && structure[y][z][x] == Blocks.AIR && world.isAirBlock(globalPos)) && !getAllowableBlocks(structure[y][z][x]).contains(new BlockMeta(block,meta))) {
```
Comment (zmaster587) [inline src/main/java/zmaster587/libVulpes/tile/multiblock/TileMultiBlock.java]: Should probably put this on multiple lines for readability
```suggestion
					if(!(structure[y][z][x] instanceof Character
					&& (Character)structure[y][z][x] == 'c')
					&& !(structure[y][z][x] instanceof Block
					&& structure[y][z][x] == Blocks.AIR
					&& world.isAirBlock(globalPos))
					&& !getAllowableBlocks(structure[y][z][x]).contains(new BlockMeta(block,meta))) {
```

File: src/main/java/zmaster587/libVulpes/tile/multiblock/TileMultiBlock.java [hunk 11]
```diff
@@ -366,10 +367,10 @@ else if(((IMultiblock)tile).getMasterBlock() == this)
...
 
 						//Can it be replaced?
-						if(block.isReplaceable(world, globalPos) && structure[y][z][x] instanceof Block && (Block)structure[y][z][x] == Blocks.AIR )
+						if(block.isReplaceable(world, globalPos) && structure[y][z][x] instanceof Block && structure[y][z][x] == Blocks.AIR )
```
Comment (zmaster587) [inline src/main/java/zmaster587/libVulpes/tile/multiblock/TileMultiBlock.java]:
```suggestion
						if(block.isReplaceable(world, globalPos)
						&& structure[y][z][x] instanceof Block
						&& structure[y][z][x] == Blocks.AIR)
```

File: src/main/java/zmaster587/libVulpes/items/ItemMaterialBlock.java (hunk 7 above)
Comment (voidsong-dragonfly) [reply src/main/java/zmaster587/libVulpes/items/ItemMaterialBlock.java]: That entire file was useless so I simply opted to remove it

File: src/main/java/zmaster587/libVulpes/tile/energy/TileCreativePowerInput.java (hunk 8 above)
Comment (voidsong-dragonfly) [reply src/main/java/zmaster587/libVulpes/tile/energy/TileCreativePowerInput.java]: Removed

**Review by zmaster587 — APPROVED** [review]

**Comment by zmaster587** [comment]
I'm going to hold off merging this in until the AR one is merged, I don't want another situation where the LibVulpes release is too new to work with the current AR release due to method changes.

**Comment by voidsong-dragonfly** [comment]
Yeah, that's a good idea

**Pull request merged**